src-control/Control/RefCount.hs

```diff
@@ -131,3 +165,231 @@ readRefCount RefCounter{countVar} = RefCount <$> readPrimVar countVar
+  deriving stock Eq
+
+newRawRefCounter :: PrimMonad m => m (RawRefCounter m)
+newRawRefCounter = do
+    countVar <- newPrimVar 1
+    return $! RawRefCounter countVar
+
+incrementRefCounter :: PrimMonad m => RawRefCounter m -> m ()
+incrementRefCounter (RawRefCounter countVar) = do
+    prevCount <- fetchAddInt countVar 1
+    assert (prevCount > 0) $ pure ()
+
+decrementRefCounter :: PrimMonad m => RawRefCounter m -> m Bool
+decrementRefCounter (RawRefCounter countVar) = do
+    prevCount <- fetchSubInt countVar 1
+    assert (prevCount > 0) $ pure $! prevCount == 1
+
+tryIncrementRefCounter :: PrimMonad m => RawRefCounter m -> m Bool
+tryIncrementRefCounter (RawRefCounter countVar) = do
+    prevCount <- atomicReadInt countVar
+    casLoop prevCount
+  where
+    -- A classic lock-free CAS loop.
+    -- Check the value before is non-zero, return failure or continue.
+    -- Atomically write the new (incremented) value if the old value is
+    -- unchanged, and return the old value (either way).
+    -- If no other thread changed the old value, we succeed.
+    -- Otherwise we go round the loop again.
+    casLoop prevCount
+      | prevCount <= 0 = return False
+      | otherwise      = do
+          prevCount' <- casInt countVar prevCount (prevCount+1)
+          if prevCount' == prevCount
+            then return True
+            else casLoop prevCount'
+
+
+-- | A reference to an object of type @a@. Use references to support prompt
+-- finalisation of object resources.
+--
+-- The reference API works relatively uniformly, both for objects that only
+-- expect to have a single reference to them, and also objects supporting
+-- reference counting that may have many references.
+--
+-- Rules of use:
+--
+-- * Each 'Ref' must eventually be released /exactly/ once with 'releaseRef'.
+-- * Use 'withRef', or 'unsafeDeRef' to (temporarily) obtain the underlying
+--   object.
+-- * After calling 'releaseRef', the operations 'withRef' and 'unsafeDeRef'
+--   must /not/ be used.
+-- * After calling 'releaseRef', any value obtained previously from
+--   'unsafeDeRef' must /not/ be used. For this reason, it is advisable to use
+--   'withRef' where possible, and be careful with use of 'unsafeDeRef'.
+--
+-- Additionally, for objects in the 'HasSharedFinaliser' class:
+--
+-- * A 'Ref' may be duplicated using 'dupRef' to produce an independent
+--   reference (which must itself be released with 'releaseRef').
+--
+-- Provided that all these rules are followed, this guarantees that the
+-- object's finaliser will be run exactly once, promptly. For objects in the
+-- 'HasSharedFinaliser' class, this is when the final reference is released.
+--
+-- In debug mode (when using CPP define @NO_IGNORE_ASSERTS@), adherence to
+-- these rules is checked dynamically.
+#ifndef NO_IGNORE_ASSERTS
+newtype Ref obj = Ref obj
+#else
+data Ref obj = Ref !obj !(IORef Bool)
+#endif
+
+#ifdef NO_IGNORE_ASSERTS
+assertNotReleased :: PrimMonad m => IORef Bool -> m ()
```

Comment:
Do we want a `HasCallStack` constraint here?

Reply:
I've added `HasCallStack` to the double-free and use-after-free assertions, and their call sites.